substratevm/src/com.oracle.svm.graal/src/com/oracle/svm/graal/isolated/IsolatedSpeculationLog.java

```diff
@@ -48,19 +53,95 @@ public void collectFailedSpeculations() {
         collectFailedSpeculations0(IsolatedCompileContext.get().getClient(), handle);
     }
 
+    private static byte[] encodeAsByteArray(SpeculationReason reason) {
+        int groupId = 0;
+        Object[] context = null;
+        try {
+            final Field groupIdField;
+            final Field contextField;
+            /*
+             * The following reflective accesses do not add runtime overhead as they are being
+             * converted to direct accesses during image building. Maintaining two separate paths
+             * that invoke {@code getDeclaredField} ensures that the analysis will be able to
+             * optimize the reflective accesses and convert them to direct ones.
+             */
+            if (reason instanceof UnencodedSpeculationReason) {
+                final Class<?> klass = UnencodedSpeculationReason.class;
+                groupIdField = klass.getDeclaredField("groupId");
+                contextField = klass.getDeclaredField("context");
+            } else {
+                final Class<?> klass = Class.forName("jdk.vm.ci.meta.EncodedSpeculationReason");
+                groupIdField = klass.getDeclaredField("groupId");
+                contextField = klass.getDeclaredField("context");
+            }
+            groupIdField.setAccessible(true);
+            groupId = groupIdField.getInt(reason);
```

Comment:
Feedback from internal review: 

> Instead of reflection, it is better to use a method handle returned by `MethodHandles.lookup().unreflectGetter` and stored in a static final field. The method handle is then really intrinsified to a direct field access.

Reply:
I tried this:
```java
if (reason instanceof UnencodedSpeculationReason) {
    final Class<?> klass = UnencodedSpeculationReason.class;
    final Field groupIdField = klass.getDeclaredField("groupId");
    final Field contextField = klass.getDeclaredField("context");
    final MethodHandle groupIdHandle = MethodHandles.lookup().unreflectGetter(groupIdField); // fails here
    final MethodHandle contextHandle = MethodHandles.lookup().unreflectGetter(contextField);
    groupId = (int) groupIdHandle.invoke(reason);
    context = (Object[]) contextHandle.invoke(reason);
} else {
    final Class<?> klass = Class.forName("jdk.vm.ci.meta.EncodedSpeculationReason");
    final Field groupIdField = klass.getDeclaredField("groupId");
    final Field contextField = klass.getDeclaredField("context");
    final MethodHandle groupIdHandle = MethodHandles.lookup().unreflectGetter(groupIdField);
    final MethodHandle contextHandle = MethodHandles.lookup().unreflectGetter(contextField);
    groupId = (int) groupIdHandle.invoke(reason);              // fails here
    context = (Object[]) contextHandle.invoke(reason);         // and here
}
```

but the if-case fails with:
```
Error: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported type java.lang.invoke.MemberName is reachable: All methods from java.lang.invoke should have been replaced during image building.
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
```
for
```java
                final MethodHandle groupIdHandle = MethodHandles.lookup().unreflectGetter(groupIdField);
```

and the else-case with:
```
Error: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Invoke with MethodHandle argument could not be reduced to at most a single call or single field access. The method handle must be a compile time constant, e.g., be loaded from a `static final` field. Method that contains the method handle invocation: java.lang.invoke.MethodHandle.invokeBasic(Object)
To enable method handles that do not require LambdaForm interpretation (e.g. because of a call to MethodHandle.bindTo()) or to diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The error is then reported at run time when the invoke is executed and the method handle has to be interpreted.
```
for
```java
                groupId = (int) groupIdHandle.invoke(reason);
                context = (Object[]) contextHandle.invoke(reason);
```
Which, if I am not mistaken, hints that the accesses were not beeing intrinsified before the change as well.

Reply:
Looks like the accesses cannot be intrinsified in the code, but what if you store them in a `static final` field? It should be possible to move the class detection logic to a `static` initializer for that.

Reply:
Tried that as well (see https://github.com/zakkak/mandrel/commit/8a6a7967324b1ab2ee73a5474b9245386b6e6dd3), but it still gives
```
Error: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Invoke with MethodHandle argument could not be reduced to at most a single call or single field access. The method handle must be a compile time constant, e.g., be loaded from a `static final` field. Method that contains the method handle invocation: java.lang.invoke.LambdaForm$MH/658839328.invoke_MT(Object, Object, Object)
```

Reply:
I finally got around to look into this today, sorry for the delay. The issue is that calling the getter on the interface type (rather than a concrete subclass which is guaranteed to have the accessed fields) introduces a type-check, which makes the method handle logic too complex to inline.
I'll merge the reflection variant.

Reply:
Thanks you @peter-hofer!
Could you please outline the process you used to figure this out (if possible with tools available to the community)?

I tried creating bgv files from building the native `native-image` but I couldn't find the graph for `encodeAsByteArray`.

Reply:
From debugging the code that creates the error in `IntrinsifyMethodHandlesInvocationPlugin`, I could gather that the method handle graph contains an `If` node where it isn't supported. I then used the command line `mx native-unittest IsolatedSpeculationLog --build-args -H:Dump=:1 -H:MethodFilter=encodeAsByteArray` to get a bgv file with the method handle graph that cannot be inlined and saw that it has a type check.
I then consulted with @christianwimmer who told me to look at the parameter node's type. That turned out to be `SpeculationReason`, which is an interface that of course doesn't have the accessed field(s), so the type check is necessary. Since we use `Class.forName` and so cannot define a statically typed variable of the subclass with the fields, we cannot proceed.

Reply:
That helps a lot, thank you for the information @peter-hofer